internal/jobworker/cgroup/cgroup.go

```diff
@@ -0,0 +1,160 @@
+	// value indicates no limit is set.
+	Memory uint64
+	// Cpus is the "cpu.max" limit applied to this cgroup. An empty value
+	// indicates no limit is set.
+	Cpus float32
+	// DiskWriteBps is the "io.max" bytes written per second limit for 8 block
+	// devices applied to this cgroup. An empty value indicates no limit is set.
+	DiskWriteBps uint64
+	// DiskReadBps is the "io.max" bytes read per second limit for 8 block
+	// devices applied to this cgroup. An empty value indicates no limit is set.
+	DiskReadBps uint64
+
+	// service is the Service a Cgroup belongs to.
+	service Service
+
+	// path is the file path to the Cgroup
+	path string
+}
+
+// CgroupOption is a function that mutates Cgroup instances. Typically used
+// with cgroups Service.CreateCgroup to create new cgroups Cgroup instances.
+type CgroupOption func(*Cgroup)
+
+// WithMemory configures a Cgroup to utilize the specified memory bytes limit.
+func WithMemory(limit uint64) CgroupOption {
+	return func(c *Cgroup) { c.Memory = limit }
+}
+
+// WithCpus configures a Cgroup to utilize the specified cpus limit.
+func WithCpus(limit float32) CgroupOption {
+	return func(c *Cgroup) { c.Cpus = limit }
+}
+
+// WithDiskWriteBps configures a Cgroup to utilize the specified bytes per
+// second limit for disk (block 8 devices) writes.
+func WithDiskWriteBps(limit uint64) CgroupOption {
+	return func(c *Cgroup) { c.DiskWriteBps = limit }
+}
+
+// WithDiskReadBps configures a Cgroup to utilize the specified bytes per
+// second limit for disk (block 8 devices) reads.
+func WithDiskReadBps(limit uint64) CgroupOption {
+	return func(c *Cgroup) { c.DiskReadBps = limit }
+}
+
+// create creates a jobworker cgroup.
+func (c Cgroup) create() error {
+	if err := os.Mkdir(c.path, fileMode); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// determine which controllers should be enabled.
+	var set []controller
+	if c.Memory > 0 {
+		set = append(set, newMemoryController(c, c.Memory))
+	}
+	if c.Cpus > 0 {
+		set = append(set, newCPUController(c, c.Cpus))
+	}
+	if c.DiskWriteBps > 0 {
+		set = append(set, newDiskWriteBpsController(c, c.DiskWriteBps))
+	}
+	if c.DiskReadBps > 0 {
+		set = append(set, newDiskReadBpsController(c, c.DiskReadBps))
+	}
+
+	for _, controller := range set {
+		if err := controller.enable(); err != nil {
+			return errors.WithStack(err)
+		}
+		if err := controller.apply(); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+// placePID adds the specified pid to the cgroup. If the pid exists in another
+// cgroup it will be moved to this cgroup.
+func (c Cgroup) placePID(pid int) error {
+	file := path.Join(c.path, cgroupProcs)
+	fd, err := os.OpenFile(file, os.O_WRONLY, fileMode)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer fd.Close()
+
+	_, err = fd.WriteString(strconv.Itoa(pid))
+	return errors.WithStack(err)
+}
+
+// remove removes the jobworker cgroup.
```

Comment:
In order to remove a cgroup, all PIDs need to have been moved out of it. This method is responsible for this removal so that cgroups may be removed when an application closes. Wouldn't not removing the cgroups leave leftover directories on the cgroup2 filesystem?

Reply:
Let's say cgroups cleanup runs on server shutdown, which mostly happens after all jobs are stopped/cancelled. That means a simple removal would work most of the time. If we do leak a directory occasionally I wouldn't cry over it, specially for the challenge. My 2c, though.